cloudinit/config/cc_ansible.py

```diff
@@ -0,0 +1,188 @@
+import logging
+import os
+import re
+import sys
+from copy import deepcopy
+from textwrap import dedent
+from typing import Optional
+
+from cloudinit.cloud import Cloud
+from cloudinit.config.schema import MetaSchema, get_meta_doc
+from cloudinit.distros import ALL_DISTROS
+from cloudinit.settings import PER_INSTANCE
+from cloudinit.subp import subp, which
+from cloudinit.util import Version, get_cfg_by_path
+
+meta: MetaSchema = {
+    "id": "cc_ansible",
+    "name": "Ansible",
+    "title": "Configure ansible for instance",
+    "frequency": PER_INSTANCE,
+    "distros": [ALL_DISTROS],
+    "activate_by_schema_keys": ["ansible"],
+    "description": dedent(
+        """\
+        This module provides ``ansible`` integration for
+        augmenting cloud-init's configuration of the local
+        node.
+
+
+        This module installs ansible during boot and
+        then uses ``ansible-pull`` to run the playbook
+        repository at the remote URL.
+        """
+    ),
+    "examples": [
+        dedent(
+            """\
+            #cloud-config
+            ansible:
+              install-method: distro
+              pull:
+                url: "https://github.com/holmanb/vmboot.git"
+                playbook-name: ubuntu.yml
+            """
+        ),
+        dedent(
+            """\
+            #cloud-config
+            ansible:
+              package-name: ansible-core
+              install-method: pip
+              pull:
+                url: "https://github.com/holmanb/vmboot.git"
+                playbook-name: ubuntu.yml
+            """
+        ),
+    ],
+}
+
+__doc__ = get_meta_doc(meta)
+LOG = logging.getLogger(__name__)
+
+
+class AnsiblePull(abc.ABC):
+    cmd_version: list = []
+    cmd_pull: list = []
+    env: dict = os.environ.copy()
+
+    def get_version(self) -> Optional[Version]:
+        stdout, _ = subp(self.cmd_version, env=self.env)
+        first_line = stdout.splitlines().pop(0)
+        matches = re.search(r"([\d\.]+)", first_line)
+        if matches:
+            version = matches.group(0)
+            return Version.from_str(version)
+        return None
+
+    def pull(self, *args) -> str:
+        stdout, _ = subp([*self.cmd_pull, *args], env=self.env)
+        return stdout
+
+    def check_deps(self):
+        if not self.is_installed():
+            raise ValueError("command: ansible is not installed")
+
+    @abc.abstractmethod
+    def is_installed(self):
+        pass
+
+    @abc.abstractmethod
+    def install(self, pkg_name: str):
+        pass
+
+
+class AnsiblePullPip(AnsiblePull):
+    def __init__(self):
+        self.cmd_pull = ["ansible-pull"]
+        self.cmd_version = ["ansible-pull", "--version"]
+        self.env["PATH"] = ":".join([self.env["PATH"], "/root/.local/bin/"])
+
+    def install(self, pkg_name: str):
+        """should cloud-init grow an interface for non-distro package
+        managers? this seems reusable
+        """
+        if not self.is_installed():
```

Comment:
@holmanb I could see `AnsiblePullPip.install` growing the smarts to install python3-pip package if not `which["pip3"]`. This can be taken to a later PR if we so desire. Maybe there are complexities to where/how  pip is installed in a system that we don't specifically want to support here?

Reply:
Agreed to carry as a todo item in next PR iterations.